### PR TITLE
do not filter kubectl get pods if -o json or yaml

### PR DESCRIPTION
**Pull request by juanvallejo**
Fixes: https://github.com/kubernetes/kubernetes/issues/38327

This patch sets the value of --show-all to true if the output format
specified is 'json' or 'yaml'.

**Release note**:
```release-note
release-note-none
```

@smarterclayton

### CODE REVIEW

**Comment by pwittrock** [comment]
@k8s-bot ok to test

File: pkg/kubectl/cmd/get.go [hunk 1]
```diff
@@ -187,7 +187,8 @@ func RunGet(f cmdutil.Factory, out, errOut io.Writer, cmd *cobra.Command, args [
 	if err != nil {
 		return err
 	}
-	if len(options.Filenames) > 0 || argsHasNames {
+	output := cmdutil.GetFlagString(cmd, "output")
+	if len(options.Filenames) > 0 || argsHasNames || output == "json" || output == "yaml" {
```
Comment (smarterclayton) [inline pkg/kubectl/cmd/get.go]: Probably would be better to default show-all to true, rather than force it to true.

Comment (smarterclayton) [inline pkg/kubectl/cmd/get.go]: I.e. kubectl get pods -o yaml --show-all=false should filter pods.

**Comment by juanvallejo** [comment]
@smarterclayton thanks for the feedback, updated the default value (and flag description) of `show-all` to `true` https://github.com/kubernetes/kubernetes/pull/39042/commits/e1ec3d80991937ce92a0eddb6844dbe487693153

**Comment by pwittrock** [comment]
@juanvallejo I think we only want to change the default if -o json or yaml is specified.  I think you should be able to check if a flag value has been specified or not (different than checking the value).  You can then keep your changes as before, but guard setting "true" on if it has not been explicitly set already.

**Comment by juanvallejo** [comment]
@pwittrock Thanks for the feedback, updated original commit to check if `show-all` flag value was set by the user before forcing its value to `true`

**Comment by pwittrock** [comment]
@k8s-bot kops aws e2e test this

**Comment by pwittrock** [comment]
@k8s-bot unit test this

**Comment by juanvallejo** [comment]
@k8s-bot bazel test this

File: pkg/kubectl/cmd/get.go [hunk 2]
```diff
@@ -187,8 +187,11 @@ func RunGet(f cmdutil.Factory, out, errOut io.Writer, cmd *cobra.Command, args [
 	if err != nil {
 		return err
 	}
-	if len(options.Filenames) > 0 || argsHasNames {
-		cmd.Flag("show-all").Value.Set("true")
+	output := cmdutil.GetFlagString(cmd, "output")
+	if len(options.Filenames) > 0 || argsHasNames || output == "json" || output == "yaml" {
```
Comment (fabianofranz) [inline pkg/kubectl/cmd/get.go]: What happens if `--output` is `go-template*` or `jsonpath*`? I believe it should apply to all too (no filtering). Also, I'd suggest to move the output check to a helper function like `OutputsRawFormat(output string) bool` or something similar.

**Comment by juanvallejo** [comment]
@fabianofranz thanks for the feedback, added a `OutputsRawFormat` helper which returns true if any machine-parsable output was specified, and also defaulted `show-all` to `true` if the `all` alias was passed

**Comment by juanvallejo** [comment]
GCI GCE flaked on https://github.com/kubernetes/kubernetes/issues/39973

**Comment by fabianofranz** [comment]
@k8s-bot gci gce e2e test this

File: pkg/kubectl/cmd/get.go [hunk 3]
```diff
@@ -182,13 +183,22 @@ func RunGet(f cmdutil.Factory, out, errOut io.Writer, cmd *cobra.Command, args [
 		return cmdutil.UsageError(cmd, usageString)
 	}
 
-	// always show resources when getting by name or filename
+	// determine if args contains "all"
+	for _, a := range args {
+		if a == "all" {
```
Comment (pwittrock) [inline pkg/kubectl/cmd/get.go]: This will match situations where resources are named 'all' 

`kubectl get pods all`

I'd either address this in a follow up for a quicker approval and submission, or fix it by matching the parsing logic for the 'all' keyword

Comment (juanvallejo) [inline pkg/kubectl/cmd/get.go]: @pwittrock thanks for the feedback. I removed this and just use the boolean value of `resource.MultipleTypesRequested`.

cc @fabianofranz this would mean that passing the `all` alias *or* requesting mixed resource types would cause the value of `show-all` to default to `true`. wdyt?

Comment (pwittrock) [inline pkg/kubectl/cmd/get.go]: ug, it looks like that code has the same issue. I am not sure if there is a great solution:

https://github.com/kubernetes/kubernetes/blob/6e5b455b7762f73f599dbec69c9ceb7cb4c8dfca/pkg/kubectl/resource/builder.go#L808

Comment (pwittrock) [inline pkg/kubectl/cmd/get.go]: Sorry if this was covered in another comment.  What is the motivation for showing all when printing all resource types?  This is unrelated to the output type.

Comment (juanvallejo) [inline pkg/kubectl/cmd/get.go]: > What is the motivation for showing all when printing all resource types? This is unrelated to the output type.

Although unrelated to the output type, I was thinking it would be a better user experience overall if `all` implied that even resources that are hidden by default would also be displayed.

> it looks like that code has the same issue. I am not sure if there is a great solution:

https://github.com/kubernetes/kubernetes/blob/6e5b455b7762f73f599dbec69c9ceb7cb4c8dfca/pkg/kubectl/resource/builder.go#L808

Would it be a valid assumption that if `all` comes before any other "kind", it can be implied that the user did mean "all resources", rather than `pod/all`? If this is the case, could we only check to see if `args[0] == "all"`, or would that be too fragile?

Comment (pwittrock) [inline pkg/kubectl/cmd/get.go]: > Although unrelated to the output type, I was thinking it would be a better user experience overall if all implied that even resources that are hidden by default would also be displayed.

This might be a lot of things, since this is showing both all resource types and hidden resources. 

> Would it be a valid assumption that if all comes before any other "kind", it can be implied that the user did mean "all resources", rather than pod/all? If this is the case, could we only check to see if args[0] == "all", or would that be too fragile?

Maybe we should check there is only 1 argument, and it is all?

```go
if len(args) == 1 && args[0] == "all"
```

Comment (juanvallejo) [inline pkg/kubectl/cmd/get.go]: > Maybe we should check there is only 1 argument, and it is all?

Sounds good to me, I can go ahead and open a PR to patch this in the `MultipleTypesRequested` helper

Comment (juanvallejo) [inline pkg/kubectl/cmd/get.go]: > This might be a lot of things, since this is showing both all resource types and hidden resources.

This makes sense, I'll remove the check for `all` from the logic for the default value of `show-all`

Comment (fabianofranz) [inline pkg/kubectl/cmd/get.go]: > Maybe we should check there is only 1 argument, and it is all?

Remember you may be interesting in getting all + something that is not in all, in which case you would have for example `kubectl get events,all`. Seems to break the proposed logic.

**Comment by pwittrock** [comment]
SGTM.  Let me know when it is updated.  (removing multiple resource types piece)

File: pkg/kubectl/cmd/get.go (hunk 3 above)
Comment (juanvallejo) [inline pkg/kubectl/cmd/get.go]: @fabianofranz 
> Remember you may be interesting in getting all + something that is not in all, in which case you would have for example kubectl get events,all. Seems to break the proposed logic.

This is a good point, I am updating the logic for `MultipleTypesRequested` in [this PR](https://github.com/kubernetes/kubernetes/pull/40046/files), and while specifying something like `$ kubectl get pods,all` does not fulfill the base case check of `if len(args) == 1 && args[0] == "all"`, the rest of the logic for that function does correctly detect multiple types in cases like this one when it splits the string by `,` and finds more than one "resource" was specified. Although this would no longer be work for this PR (if we had just decided to check for "all" using `len(args) == 1 && args[0] == "all"` rather than calling the entire `MultipleTypesRequested` helper)

**Comment by juanvallejo** [comment]
@pwittrock 
> SGTM. Let me know when it is updated. (removing multiple resource types piece)

Done!

**Comment by smarterclayton** [comment]
After looking at this and staring at it for a bit while fixing other bugs, I think I was wrong to want this.  It makes behavior of get inconsistent and is surprising.  Users should learn to use `-a`.

I'm going to open a PR to revert this behavior for 1.6 unless anyone objects.

**Comment by smarterclayton** [comment]
Since this has CLI API implications, I'd prefer to make a decision for 1.6 so we don't have to live with it.

**Comment by smarterclayton** [comment]
Instead, we should simply ensure we don't show the message for filtering for generic output.
